### PR TITLE
⭐ new `transfer.py` module

### PR DESCRIPTION
**Pull request by selipot**
This is the start of a draft PR to implement the `transfer` module for the wind driven models of Lilly and Elipot 2021 and Elipot and Gille 2009.

### CODE REVIEW

File: clouddrift/transfer.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    zo = np.divide(delta**2, mu)
+    s = np.sign(coriolis_frequency) * np.sign(1 + omega / coriolis_frequency)
+
+    xiz, zih, xi0 = _xis(s, zo, delta, z, omega, coriolis_frequency, bld)
+
+    if bld == np.inf:
+        if mu == 0:
+            # Ekman solution
+            print("Ekman solution")
```
Comment (kevinsantana11) [inline clouddrift/transfer.py:103]: We should use the `logging` facility provided by python for printing log messages or we should remove this if it was just for testing purposes

Comment (selipot) [reply clouddrift/transfer.py:103]: It will be removed yes. Using this for development.

**Comment by selipot** [comment]
More tests are probably needed but it is in good shape.

File: clouddrift/transfer.py [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+        denom = i1h * k10 - i10 * k1h
+        G = coeff * numer / denom
+
+    return G
+
+
+def bessels_freeslip(
+    xiz: Union[float, np.ndarray],
+    xih: Union[float, np.ndarray],
+    xi0: Union[float, np.ndarray, None] = None,
```
Comment (philippemiron) [inline clouddrift/transfer.py:658]: Add `Optional` here too.

File: clouddrift/transfer.py [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+        i10 = iv(1, xi0)
+        k0z = kv(0, xiz)
+        numer = k1h * i0z + i1h * k0z
+        denom = i1h * k10 - i10 * k1h
+        G = coeff * numer / denom
+
+    return G
+
+
+def bessels_freeslip(
```
Comment (philippemiron) [inline clouddrift/transfer.py:655]: Do we want those part of the API of the modules? Or they should start with `_`.

File: clouddrift/transfer.py [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+        return k0z, i0z, k0h, i0h, k10, i10
+    else:
+        return k0z, i0z, k0h, i0h, np.nan, np.nan
+
+
+def besseltildes_noslip(
+    xiz: Union[float, np.ndarray],
+    xih: Union[float, np.ndarray],
+    xi0: Union[float, np.ndarray],
+    nterms=30,
```
Comment (philippemiron) [inline clouddrift/transfer.py:709]:
```suggestion
    nterms: Optional[int] = 30,
```

File: clouddrift/transfer.py [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+    return k0z, i0z, k0h, i0h, k10, i10
+
+
+def kvtilde(
+    nu: int,
+    z: Union[float, np.ndarray],
+    nterms=30,
+) -> Union[float, np.ndarray]:
+    """
+    Compute the 30-term expansion about the large-argument exponential behavior of the Bessel functions
```
Comment (philippemiron) [inline clouddrift/transfer.py:737]:
```suggestion
    Compute the nterm expansion about the large-argument exponential behavior of the Bessel functions
```

File: clouddrift/transfer.py [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+    k10 = kvtilde(1, xi0, nterms)
+    i10 = ivtilde(1, xi0, nterms)
+
+    return k0z, i0z, k0h, i0h, k10, i10
+
+
+def kvtilde(
+    nu: int,
+    z: Union[float, np.ndarray],
+    nterms=30,
```
Comment (philippemiron) [inline clouddrift/transfer.py:734]:
```suggestion
    nterms: Optional[int] = 30,
```

File: clouddrift/transfer.py [hunk 7]
```diff
@@ -0,0 +1,843 @@
...
+    K = sqrt(np.pi / (2 * z)) * (np.dot(1.0 / zk, ak))
+    K = K.reshape(sizez)
+
+    return K
+
+
+def ivtilde(
+    nu: int,
+    z: Union[float, np.ndarray],
+    nterms=30,
```
Comment (philippemiron) [inline clouddrift/transfer.py:770]:
```suggestion
    nterms: Optional[int] = 30,
```

File: clouddrift/transfer.py [hunk 8]
```diff
@@ -0,0 +1,843 @@
...
+) -> Union[float, np.ndarray]:
+    z = np.asarray(
```
Comment (philippemiron) [inline clouddrift/transfer.py:772]: Docstring

File: clouddrift/transfer.py [hunk 9]
```diff
@@ -0,0 +1,843 @@
...
+        * sqrt(2)
+        * _rot(s * np.pi / 4)
+        * np.divide(zo, delta)
+        * sqrt(np.abs(1 + omega / coriolis_frequency))
+    )
+
+    return xiz, xih, xi0
+
+
+def _rot(x) -> complex:
```
Comment (philippemiron) [inline clouddrift/transfer.py:839]: Typing of x.

File: clouddrift/transfer.py [hunk 10]
```diff
@@ -0,0 +1,843 @@
...
+    xih: Union[float, np.ndarray],
+    xi0: Union[float, np.ndarray, None] = None,
+) -> Tuple[
+    Union[float, np.ndarray],
+    Union[float, np.ndarray],
+    Union[float, np.ndarray],
+    Union[float, np.ndarray],
+    Union[float, np.ndarray],
+    Union[float, np.ndarray],
+]:
```
Comment (philippemiron) [inline clouddrift/transfer.py:691]: Docstring

File: clouddrift/transfer.py [hunk 11]
```diff
@@ -0,0 +1,843 @@
...
```
Comment (philippemiron) [inline clouddrift/transfer.py:666]: Docstring

File: tests/transfer_tests.py [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+                test_name = "kvTILDE for z with phase of -pi/4"
+
+            with self.subTest(test_name=test_name):
+                self.assertTrue(
+                    np.allclose(np.abs((bk0 - bk) / bk), 0, atol=atol),
+                    msg=f"Failed: {test_name}",
+                )
+
+            for i in range(2):
+                # pass
```
Comment (philippemiron) [inline tests/transfer_tests.py:302]: I guess this can be removed.

File: clouddrift/transfer.py [hunk 13]
```diff
@@ -0,0 +1,843 @@
...
+            )
+
+    # set G to nan where z > bld; may be mathematcially valid but not physically meaningful
+    G[z_grid > bld] = np.nan
+
+    # analytical gradients of the transfer function for mu = 0 and free slip, lilly method
+    if boundary_condition == "no-slip" and method == "lilly" and mu == 0:
+        s = np.sign(cor_freq_) * np.sign(1 + omega_grid / cor_freq_)
+        Gamma = (
+            sqrt(2)
```
[the remark this answers is not in the document]
Comment (selipot) [reply clouddrift/transfer.py:163]: For positive number nothing different I believe. For complex and negative numbers it handles things properly. I thought easier to use it for everything in this module. Does this work?

File: clouddrift/transfer.py (hunk 3 above)
Comment (selipot) [reply clouddrift/transfer.py:655]: they should indeed.

**Comment by selipot** [comment]
For this PR I want to use the type declaration/typing using python 3.10 conventions ( [PEP 604](https://peps.python.org/pep-0604/) and [PEP 585](https://peps.python.org/pep-0585/)). @kevinsantana11 should we have another PR before that to remove the test suite for Python 3.9?

**Comment by philippemiron** [comment]
> For this PR I want to use the type declaration/typing using python 3.10 conventions ( [PEP 604](https://peps.python.org/pep-0604/) and [PEP 585](https://peps.python.org/pep-0585/)). @kevinsantana11 should we have another PR before that to remove the test suite for Python 3.9? As part of this PR, you already adjusted the typing for 3.9, right? 

I agree that PEP 604, in particular, is much nicer and simplifies the syntax a lot, but I would suggest we merge it and then do a single PR to fix all the types.

**Comment by selipot** [comment]
> > For this PR I want to use the type declaration/typing using python 3.10 conventions ( [PEP 604](https://peps.python.org/pep-0604/) and [PEP 585](https://peps.python.org/pep-0585/)). @kevinsantana11 should we have another PR before that to remove the test suite for Python 3.9?
> 
> I agree that PEP 604, in particular, is much nicer and simplifies the syntax a lot, but I would suggest we merge it and then do a single PR to fix all the types.
> 
> It looks like you already made the modification to support 3.9, correct?

Right now I believe the way I wrote the code is being supported by 3.9. So yes we can do a later PR. Hopefully with some shell scripting tool like sed to automatically replace all the typing.

**Comment by philippemiron** [comment]
`pyupgrade` should do it!

https://adamj.eu/tech/2021/05/21/python-type-hints-how-to-upgrade-syntax-with-pyupgrade/

**Comment by philippemiron** [comment]
Is it normal that it doesn't pass the test now?

**Comment by selipot** [comment]
> Is it normal that it doesn't pass the test now?

Yes, working on this. It is a draft and there are still some problems.

**Comment by selipot** [comment]
@kevinsantana11 I think we should merge, what do you think?

**Comment by selipot** [comment]
I'll start a demo repo notebook next

**Comment by selipot** [comment]
@kevinsantana11 can we merge?

**Comment by kevinsantana11** [comment]
@selipot yes, please go ahead and merge

**Pull request merged**